testsuite/shared/src/main/java/org/wildfly/test/stability/StabilityServerSetupSnapshotRestoreTasks.java

```diff
@@ -0,0 +1,193 @@
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_OPERATION_DESCRIPTION_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_OPERATION_NAMES_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.RELOAD_ENHANCED;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.STABILITY;
+import static org.jboss.as.server.controller.descriptions.ServerDescriptionConstants.SERVER_ENVIRONMENT;
+
+/**
+ * For tests that need to run under a specific server stability level,
+ * the server setup tasks from the inner classes can be used to change the stability level of the server to the desired level.
+ * Once the test is done, the original stability level is restored.
+ *
+ * In order to not pollute the configuration with XML from a different stability level following the run of the test,
+ * it takes a snapshot of the server configuration in the setup() method, and
+ */
+public abstract class StabilityServerSetupSnapshotRestoreTasks implements ServerSetupTask {
+
+    private final Stability desiredStability;
+    private volatile Stability originalStability;
+
+    private AutoCloseable snapshot;
+
+
+    public StabilityServerSetupSnapshotRestoreTasks(Stability desiredStability) {
+        this.desiredStability = desiredStability;
+    }
+
+    @Override
+    public final void setup(ManagementClient managementClient) throws Exception {
+        // Make sure the desired stability level is one of the ones supported by the server
+        Set<Stability> supportedStabilityLevels = getSupportedStabilityLevels();
+        Assume.assumeTrue(
+                String.format("%s is not a supported stability level", desiredStability, supportedStabilityLevels),
+                supportedStabilityLevels.contains(desiredStability));
+
+        // Check the reload-enhanced operation exists in the current stability level
+        Assume.assumeTrue(
+                "The reload-enhanced operation is not registered at this stability level",
+                checkReloadEnhancedOperationIsAvailable(managementClient));
+
+        // Check the reload-enhanced operation exists in the stability level we want to load to so that
+        // we can reload back to the current one
+        Stability reloadOpStability = getReloadEnhancedOperationStabilityLevel(managementClient);
+        Assume.assumeTrue(desiredStability.enables(reloadOpStability));
+
+
+
+        originalStability = readCurrentStability(managementClient.getControllerClient());
+
+        // Take a snapshot, indicating that when reloading we want to go back to the original stability
+        snapshot = ServerSnapshot.takeSnapshot(managementClient, originalStability);
+
+        // All good, let's do it!
+        reloadToDesiredStability(managementClient.getControllerClient(), desiredStability);
+
+        // Do any additional setup from the sub-classes
+        doSetup(managementClient);
```

Comment:
I suggest dropping this.

The ServerSetup annotation supports task chaining, which seems like a better programming idiom than subclassing for tests that need to do tasks, and will continue to need to do those tasks as the feature being tested changes stability levels.

Reply:
Hmm, but you've combined the snapshot stuff into this, which seems orthogonal and forces users to follow that paradigm.

We need tasks that just change the runtime stability level and don't mess with the persistent config.

Reply:
I mean this is ok, but we also need basic tasks that just reload.

Reply:
Just for the record, as discussed in chat, if changes are made to the config while the server is at a lower stability level the resulting configuration will not work on the reloaded server. So we are keeping it as-is for now

Reply:
Yep this is ok. 

I won't hit the 'Resolve conversation' button because this may be a useful conversation for future maintainers.